engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -200,6 +202,44 @@ public void registerPolyglotSymbol(String name, Object sym) {
     polyglotSymbols.put(name, sym);
   }
 
+  private final Map<String, Map<String, Supplier<Function>>> polyMethods = new HashMap<>();
```

Comment:
Honestly, the code I have put into `ModuleScope` is so ugly that it deserves a _total rewrite_. With e279793 we seem to have the desirable functionality, but the implementation has to improve significantly.

Pavel, do you think the following would be a fine goal?

**Give every Java class an Enso `Type`**

I am still a bit afraid of doing so, but it would solve a lot of problems. If there was a _Type wrapper around each Java class_ used in Enso, we could easily assign the extension methods the _regular way_ - e.g. without duplicating all these internal fields and register methods. We would just assign them to the appropriate `Type` wrapping the Java class. Moreover Java types would appear in the IDE as types. Something that
- #8009

wanted to achieve a long time ago.

I would just need to _open_ `Type`  _up_ and a properly "cache" the type and java class mapping - guarantee it is 1:1. We probably still need the _eigen type_ concept (even for Java classes) as that's where Enso stores static methods. The current implementation struggles a lot differentiating between static and non-static methods (see current [MethodInvokeNode](https://github.com/enso-org/enso/pull/9129/files#diff-6bc7c2f8a27bdb61f738a75b6accb867d8b5336e3f7d1ea3c7128b26f36e2fa6R852)). And yes, re-implementing #3764 to register each method only once and _just dispatch properly_ would make me happier, but it is another thing I am hesitating/afraid to do...